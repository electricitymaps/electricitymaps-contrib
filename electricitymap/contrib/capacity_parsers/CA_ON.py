import logging
from datetime import datetime
from typing import Any

import pandas as pd
from requests import Response, Session

from electricitymap.contrib.config import ZoneKey

logger = logging.getLogger(__name__)

MODE_MAPPING = {
    "Nuclear": "nuclear",
    "Hydroelectric": "hydro",
    "Gas/Oil": "gas",
    "Wind": "wind",
    "Biofuel": "biomass",
    "Solar": "solar",
}


"""Disclaimer: only valid for real-time data, historical capacity is not available"""
OUTLOOK_URL = "https://www.ieso.ca/-/media/Files/IESO/Document-Library/planning-forecasts/reliability-outlook/ReliabilityOutlookTables_{date}.ashx"


def get_data_from_url(session: Session, target_datetime: datetime) -> pd.DataFrame:
    file_url = OUTLOOK_URL.format(date=target_datetime.strftime("%Y%b"))
    file_response: Response = session.get(file_url)
    if "Error-404" not in file_response.url and file_response.ok:
        df = pd.read_excel(
            file_response.url, sheet_name="Table 4.1", header=4, skipfooter=3
        )
        return df
    else:
        raise ValueError(
            f"Failed to fetch capacity data for IESO from url: {target_datetime.date()}"
        )


def fetch_production_capacity(
    zone_key: ZoneKey, target_datetime: datetime, session: Session
) -> dict[str, Any]:
    df = get_data_from_url(session, target_datetime)

    df = df.rename(
        columns={"Fuel Type": "mode", "Total Installed Capacity\n(MW)": "value"}
    )
    df = df[["mode", "value"]]
    df["mode"] = df["mode"].apply(lambda x: MODE_MAPPING[x])
    capacity = {}
    for _idx, data in df.iterrows():
        capacity[data["mode"]] = {
            "datetime": target_datetime.strftime("%Y-%m-%d"),
            "value": round(data["value"], 0),
            "source": "ieso.ca",
        }
    logger.info(
        f"Fetched capacity for {zone_key} on {target_datetime.date()}: \n{capacity}"
    )
    return capacity


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    fetch_production_capacity(ZoneKey("CA_ON"), datetime(2023, 3, 1), Session())
