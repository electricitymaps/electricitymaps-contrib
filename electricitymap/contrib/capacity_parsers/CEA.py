from datetime import datetime
from logging import getLogger
from typing import Any

import pandas as pd
from requests import Response, Session

from electricitymap.contrib.config.capacity import CAPACITY_PARSER_SOURCE_TO_ZONES
from electricitymap.contrib.lib.types import ZoneKey

logger = getLogger(__name__)
SOURCE = "cea.nic.in"
CAPACITY_URL = "https://cea.nic.in/api/installed_capacity.php"
REGION_MAPPING = {
    "Northern": "IN-NO",
    "Western": "IN-WE",
    "Southern": "IN-SO",
    "Eastern": "IN-EA",
    "North Eastern": "IN-NE",
}
MODE_MAPPING = {
    "Coal": "coal",
    "Gas": "gas",
    "Diesel": "oil",
    "Nuclear": "nuclear",
    "Hydro": "hydro",
}
IN_ZONES = CAPACITY_PARSER_SOURCE_TO_ZONES["CEA"]

IN_STATE_TO_ZONE_MAPPING = {
    "Andhra Pradesh": "IN-SO",
    "Arunachal Pradesh": "IN-NE",
    "Assam": "IN-NE",
    "Bihar": "IN-EA",
    "Chhatisgarh": "IN-WE",
    "Goa": "IN-WE",
    "Gujarat": "IN-WE",
    "Haryana": "IN-NO",
    "Himachal Pradesh": "IN-NO",
    "Jammu & Kashmir": "IN-NO",
    "Jharkhand": "IN-EA",
    "Karnataka": "IN-SO",
    "Kerala": "IN-SO",
    "Ladakh": "IN-NO",
    "Madhya Pradesh": "IN-WE",
    "Maharashtra": "IN-WE",
    "Manipur": "IN-NE",
    "Meghalaya": "IN-NE",
    "Mizoram": "IN-NE",
    "Nagaland": "IN-NE",
    "Odisha": "IN-EA",
    "Punjab": "IN-NO",
    "Rajasthan": "IN-NO",
    "Sikkim": "IN-EA",
    "Tamil Nadu": "IN-SO",
    "Telangana": "IN-SO",
    "Tripura": "IN-NE",
    "Uttar Pradesh": "IN-NO",
    "Uttarakhand": "IN-NO",
    "West Bengal": "IN-EA",
    "Delhi": "IN-NO",
    "Pondicherry": "IN-SO",
}

MNRE_MODE_MAPPING = {
    "Small Hydro Power": "hydro",
    "Wind Power": "wind",
    "Bio Power Total": "biomass",
    "Solar Power Total": "solar",
}


def fetch_production_capacity_for_all_zones(
    target_datetime: datetime, session: Session
) -> dict[str, Any] | None:
    logger.warning(
        "Renewable capacity is not available and should be downloaded from https://mnre.gov.in/physical-progress/"
    )
    r: Response = session.get(CAPACITY_URL)
    data = r.json()
    df = pd.DataFrame(data)

    df_filtered = df.loc[df["Month"] == target_datetime.strftime("%b-%Y")].copy()
    if df_filtered.empty:
        raise logger.warning(f"No capacity data for IN zones in {target_datetime.year}")
    df_filtered["datetime"] = pd.to_datetime(df_filtered["Month"])
    df_filtered["zoneKey"] = df_filtered["Region"].map(REGION_MAPPING)
    df_filtered = df_filtered[["datetime", "zoneKey"] + list(MODE_MAPPING.keys())]

    df_filtered = df_filtered.set_index(["datetime", "zoneKey"])
    all_capacity = {}
    for idx, data in df_filtered.iterrows():
        capacity_dict = {}
        for cea_mode in data.index:
            mode = MODE_MAPPING[cea_mode]
            capacity_dict[mode] = {
                "datetime": idx[0].strftime("%Y-%m-%d"),
                "value": float(data[cea_mode]),
                "source": SOURCE,
            }
        all_capacity[idx[1]] = capacity_dict
    return all_capacity


def fetch_production_capacity(
    target_datetime: datetime, zone_key: ZoneKey, session: Session
) -> dict[str, Any] | None:
    all_capacities = fetch_production_capacity_for_all_zones(target_datetime, session)
    return all_capacities[zone_key]


# manual update, the output of the function should be added to the config file
def get_renewable_capacity(path: str, zone_key: ZoneKey | None = None) -> None:
    """Extract renewable capacity from MNRE report which then should be added to the config file"""
    df = pd.read_excel(path, skiprows=2)
    for col in df.columns:
        if "Unnamed:" in col:
            df = df.rename(columns={col: df[col].iloc[0]})
    df = df.iloc[2:]
    cols_to_keep = [
        "STATES / Uts",
        "Small Hydro Power",
        "Wind Power",
        "Bio Power Total",
        "Solar Power Total",
    ]
    df_filtered = df[cols_to_keep].copy()
    df_filtered = df_filtered.rename(
        columns={**{"STATES / Uts": "zoneKey"}, **MNRE_MODE_MAPPING}
    )
    df_filtered["zoneKey"] = df_filtered["zoneKey"].map(IN_STATE_TO_ZONE_MAPPING)
    df_filtered = df_filtered.groupby(["zoneKey"]).sum()
    all_capacity = {}
    for idx, data in df_filtered.iterrows():
        capacity_dict = {}
        for mode in data.index.sort_values():
            capacity_dict[mode] = round(data[mode], 1)
        all_capacity[idx] = capacity_dict
    if zone_key:
        return all_capacity[zone_key]
    return all_capacity


if __name__ == "__main__":
    print(
        get_renewable_capacity(
            "/Users/mathildedaugy/Downloads/202401091363179073.xlsx", "IN-SO"
        )
    )
