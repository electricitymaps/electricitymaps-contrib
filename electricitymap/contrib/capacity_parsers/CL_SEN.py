import logging
from datetime import datetime
from typing import Any

import pandas as pd
from bs4 import BeautifulSoup
from requests import Response, Session

from electricitymap.contrib.config import ZoneKey

logger = logging.getLogger(__name__)

# Mapping conventional thermal as unknown as the production parser data is aggregated
MODE_MAPPING = {
    "Hídrico": "hydro",
    "Carbón": "unknown",
    "Diésel": "unknown",
    "Gas Natural": "unknown",
    "Eólico": "wind",
    "Solar": "solar",
    "Termosolar": "solar",
    "Geotérmico": "geothermal",
    "Otros*": "unknown",
}

SOURCE = "coordinador.cl"


def fetch_production_capacity(
    zone_key: ZoneKey, target_datetime: datetime, session: Session
) -> dict[str, Any] | None:
    url = "https://www.coordinador.cl/reportes-y-estadisticas/#Estadisticas"
    r: Response = session.get(url)
    soup = BeautifulSoup(r.text, "html.parser")

    capacity_link = soup.find_all(
        "a", href=lambda x: x and "hist_cap_inst_por_tecnologia" in x
    )
    if not len(capacity_link):
        logger.error(
            f"{zone_key}: No capacity data available for year {target_datetime.year}"
        )

    df = pd.read_excel(
        capacity_link[0].get("href"),
        sheet_name="Capacidad por Tecnología",
        header=2,
        skipfooter=2,
    )
    df = df.drop(columns=["Unnamed: 0", "TOTAL"])
    df = df.rename(columns={"Año": "datetime"})
    df = df.melt(id_vars=["datetime"], var_name="mode", value_name="value")
    df["mode"] = df["mode"].apply(lambda x: MODE_MAPPING[x.strip()])

    df = df.groupby(["datetime", "mode"])[["value"]].sum().reset_index()

    if target_datetime.year in df["datetime"].unique():
        df = df.loc[df["datetime"] == target_datetime.year]
        capacity = {}
        for _idx, data in df.iterrows():
            mode_capacity = {}
            mode_capacity["datetime"] = target_datetime.strftime("%Y-%m-%d")
            mode_capacity["value"] = round(data["value"], 0)
            mode_capacity["source"] = SOURCE
            capacity[data["mode"]] = mode_capacity
        logger.info(
            f"Fetched capacity for {zone_key} on {target_datetime.date()}: \n {capacity}"
        )
        return capacity
    else:
        logger.error(
            f"{zone_key}: No capacity data available for year {target_datetime.year}"
        )


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    fetch_production_capacity("CL-SEN", datetime(2022, 1, 1), Session())
