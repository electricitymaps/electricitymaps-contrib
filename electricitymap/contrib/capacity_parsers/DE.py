import logging
from datetime import datetime
from typing import Any

from requests import Response, Session

from electricitymap.contrib.config import ZoneKey

logger = logging.getLogger(__name__)

MODE_MAPPING = {
    "Nuclear": "nuclear",
    "Fossil brown coal / lignite": "coal",
    "Fossil hard coal": "coal",
    "Fossil gas": "gas",
    "Fossil oil": "oil",
    "Other, non-renewable": "unknown",
    "Hydro Run-of-River": "hydro",
    "Hydro pumped storage": "hydro storage",
    "Battery Storage (Power)": "battery storage",
    "Biomass": "biomass",
    "Wind onshore": "wind",
    "Wind offshore": "wind",
    "Solar": "solar",
}

SOURCE = "Fraunhofer ISE"

REQUEST_URL = "https://api.energy-charts.info/installed_power?country=de&time_step=yearly&installation_decommission=false"
CONVERT_GW_TO_MW = 1000


def convert_to_mw(value: float | int | None) -> float:
    if isinstance(value, int | float):
        return float(value) * CONVERT_GW_TO_MW
    else:
        return None


def fetch_production_capacity(
    zone_key: ZoneKey, target_datetime: datetime, session: Session
) -> dict[str, Any] | None:
    r: Response = session.get(REQUEST_URL)
    if not r.ok:
        raise ValueError(
            f"Failed to fetch capacity data for DE at {target_datetime.date()}"
        )
    data = r.json()
    all_capacity = {}
    for i in range(len(data["production_types"])):
        capacity_mode = {}
        mode = data["production_types"][i]["name"]
        for j in range(len(data["time"])):
            capacity_mode[int(data["time"][j])] = convert_to_mw(
                data["production_types"][i]["data"][j]
            )

        all_capacity[mode] = capacity_mode

    capacity = update_capacity_breakdown(all_capacity, target_datetime)

    logger.info(
        f"Fetched capacity for {zone_key} on {target_datetime.date()}: \n {capacity}"
    )
    return capacity


def update_capacity_breakdown(
    all_capacity: dict[str, Any], target_datetime: datetime
) -> dict[str, Any]:
    capacity = {}

    for mode, capacity_data in all_capacity.items():
        mapped_mode = MODE_MAPPING.get(mode)
        if mode not in MODE_MAPPING:
            continue
        if mapped_mode in capacity:
            capacity[mapped_mode]["value"] += round(
                capacity_data[target_datetime.year], 0
            )
        else:
            capacity[mapped_mode] = {
                "datetime": target_datetime.strftime("%Y-%m-%d"),
                "value": round(capacity_data[target_datetime.year], 0),
                "source": SOURCE,
            }

    return capacity


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    fetch_production_capacity(ZoneKey("DE"), datetime(2023, 1, 1), Session())
