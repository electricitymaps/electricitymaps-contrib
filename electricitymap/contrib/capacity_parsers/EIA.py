import calendar
from datetime import datetime
from logging import getLogger
from typing import Any

import pandas as pd
from requests import Response, Session

from electricitymap.contrib.config import ZoneKey
from electricitymap.contrib.config.capacity import CAPACITY_PARSER_SOURCE_TO_ZONES
from electricitymap.contrib.config.constants import PRODUCTION_MODES
from parsers.EIA import REGIONS
from parsers.lib.utils import get_token

logger = getLogger(__name__)

CAPACITY_URL = "https://api.eia.gov/v2/electricity/operating-generator-capacity/data/?frequency=monthly&data[0]=nameplate-capacity-mw&facets[balancing_authority_code][]={}"
SOURCE = "EIA.gov"
US_ZONES = CAPACITY_PARSER_SOURCE_TO_ZONES["EIA"]
TECHNOLOGY_TO_MODE = {
    "All Other": "unknown",
    "Batteries": "battery storage",
    "Coal Integrated Gasification Combined Cycle": "coal",
    "Conventional Hydroelectric": "hydro",
    "Conventional Steam Coal": "coal",
    "Flywheels": "battery storage",
    "Geothermal": "geothermal",
    "Hydroelectric Pumped Storage": "hydro storage",
    "Landfill Gas": "biomass",
    "Municipal Solid Waste": "biomass",  # or unknown?
    "Natural Gas Fired Combined Cycle": "gas",
    "Natural Gas Fired Combustion Turbine": "gas",
    "Natural Gas Internal Combustion Engine": "gas",
    "Natural Gas Steam Turbine": "gas",
    "Natural Gas with Compressed Air Storage": "gas",
    "Nuclear": "nuclear",
    "Offshore Wind Turbine": "wind",
    "Onshore Wind Turbine": "wind",
    "Other Gases": "gas",
    "Other Natural Gas": "gas",
    "Other Waste Biomass": "biomass",
    "Petroleum Coke": "oil",
    "Petroleum Liquids": "oil",
    "Solar Photovoltaic": "solar",
    "Solar Thermal with Energy Storage": "solar",
    "Solar Thermal without Energy Storage": "solar",
    "Wood/Wood Waste Biomass": "biomass",
}
CAPACITY_MODES = PRODUCTION_MODES + ["hydro storage", "battery storage"]


def format_capacity(df: pd.DataFrame, target_datetime: datetime) -> dict[str, Any]:
    df = df.copy()
    df = df.loc[df["statusDescription"] == "Operating"]
    df["mode"] = df["technology"].map(TECHNOLOGY_TO_MODE)
    df["nameplate-capacity-mw"] = pd.to_numeric(
        df["nameplate-capacity-mw"], errors="coerce"
    )
    df_aggregated = df.groupby(["mode"])[["nameplate-capacity-mw"]].sum().reset_index()
    capacity_dict = {}
    for mode in CAPACITY_MODES:
        mode_dict = {}
        mode_dict["value"] = round(
            float(
                df_aggregated.loc[df_aggregated["mode"] == mode][
                    "nameplate-capacity-mw"
                ].sum()
            ),
            1,
        )
        mode_dict["source"] = SOURCE
        mode_dict["datetime"] = target_datetime.strftime("%Y-%m-%d")
        capacity_dict[mode] = mode_dict
    return capacity_dict


def fetch_production_capacity(
    zone_key: ZoneKey,
    target_datetime: datetime,
    session: Session,
) -> dict[str, Any] | None:
    API_KEY = get_token("EIA_KEY")
    url_prefix = CAPACITY_URL.format(REGIONS[zone_key])
    start_date = target_datetime.strftime("%Y-%m-01")
    end_date = target_datetime.replace(
        day=calendar.monthrange(target_datetime.year, target_datetime.month)[1]
    ).strftime("%Y-%m-%d")
    url = f"{url_prefix}&api_key={API_KEY}&start={start_date}&end={end_date}&sort[0][column]=period&sort[0][direction]=desc&offset=0&length=5000"
    r: Response = session.get(url)
    json_data = r.json()

    if json_data.get("response", {}).get("data", []) != []:
        data = pd.DataFrame(json_data["response"]["data"])
        capacity_dict = format_capacity(data, target_datetime)
        logger.info(
            f"Fetched capacity data for {zone_key} at {target_datetime.strftime('%Y-%m')}: \n{capacity_dict}"
        )
        return capacity_dict
    else:
        logger.warning(
            f"Failed to fetch capacity data for {zone_key} at {target_datetime.strftime('%Y-%m')}"
        )


def fetch_production_capacity_for_all_zones(
    target_datetime: datetime, session: Session | None = None
) -> dict[str, Any]:
    eia_capacity = {}
    if session is None:
        session = Session()
    for zone in US_ZONES:
        try:
            zone_capacity = fetch_production_capacity(zone, target_datetime, session)
            if zone_capacity:
                eia_capacity[zone] = zone_capacity
        except Exception as e:
            logger.error(
                f"Error fetching production capacity for {zone} at {target_datetime.strftime('%Y-%m')}: {e}"
            )
    return eia_capacity
