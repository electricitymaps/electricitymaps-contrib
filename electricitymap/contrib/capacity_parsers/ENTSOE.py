from datetime import datetime
from logging import getLogger
from typing import Any

from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.config import ZoneKey
from electricitymap.contrib.config.capacity import CAPACITY_PARSER_SOURCE_TO_ZONES
from parsers.ENTSOE import (
    ENTSOE_DOMAIN_MAPPINGS,
    ENTSOE_PARAMETER_BY_GROUP,
    query_ENTSOE,
)

"""
Update capacity configurations for ENTOS-E zones for a chosen year.
"""
logger = getLogger(__name__)
SOURCE = "entsoe.eu"

ENDPOINT = "/api"
ENTSOE_HOST = "https://web-api.tp.entsoe.eu"


EU_PROXY = "https://eu-proxy-jfnx5klx2a-ew.a.run.app{endpoint}?host={host}"

ENTSOE_ENDPOINT = ENTSOE_HOST + ENDPOINT
ENTSOE_EU_PROXY_ENDPOINT = EU_PROXY.format(endpoint=ENDPOINT, host=ENTSOE_HOST)


ENTSOE_ZONES = CAPACITY_PARSER_SOURCE_TO_ZONES["ENTSOE"]

# ENTSOE does not have battery storage capacity and the data needs to be collected from other sources for the following zones
# TODO monitor this list and update if necessary
ZONES_WITH_BATTERY_STORAGE = ["FR"]


# reallocate B10 to hydro storage
ENTSOE_CODE_TO_EM_MAPPING = ENTSOE_PARAMETER_BY_GROUP.copy()
ENTSOE_CODE_TO_EM_MAPPING.update({"B10": "hydro storage", "B25": "battery storage"})


def query_capacity(
    in_domain: str, session: Session, target_datetime: datetime
) -> str | None:
    params = {
        "documentType": "A68",
        "processType": "A33",
        "in_Domain": in_domain,
    }
    return query_ENTSOE(
        session,
        params,
        target_datetime=target_datetime,
        span=(0, 72),  # DO NOT USE A NEGATIVE LOOKBACK
        function_name=query_capacity.__name__,
    )


def fetch_production_capacity(
    zone_key: ZoneKey, target_datetime: datetime, session: Session
) -> dict[str, Any] | None:
    xml_str = query_capacity(ENTSOE_DOMAIN_MAPPINGS[zone_key], session, target_datetime)
    soup = BeautifulSoup(xml_str, "html.parser")
    # Each time series is dedicated to a different fuel type.
    capacity_dict = {}
    for timeseries in soup.find_all("timeseries"):
        fuel_code = str(
            timeseries.find_all("mktpsrtype")[0].find_all("psrtype")[0].contents[0]
        )
        point = timeseries.find_all("point")
        value = float(point[0].find_all("quantity")[0].contents[0])
        if ENTSOE_CODE_TO_EM_MAPPING[fuel_code] not in capacity_dict:
            capacity_dict[ENTSOE_CODE_TO_EM_MAPPING[fuel_code]] = {
                "value": 0,
                "datetime": f"{target_datetime.year}-01-01",
                "source": SOURCE,
            }
        capacity_dict[ENTSOE_CODE_TO_EM_MAPPING[fuel_code]]["value"] += value
    if capacity_dict:
        logger.info(
            f"Capacity data for {zone_key} on {target_datetime.date()}: \n{capacity_dict}"
        )
        if zone_key in ZONES_WITH_BATTERY_STORAGE:
            logger.info(
                f"\n\n Warning: {zone_key} has battery storage, data source can be found on the contrib wiki \n\n"
            )
        return capacity_dict
    else:
        logger.warning(
            f"Failed to fecth capacity data for {zone_key} on {target_datetime.date()}"
        )


def fetch_production_capacity_for_all_zones(
    target_datetime: datetime, session: Session | None = None
) -> dict[str, Any]:
    capacity_dict = {}
    if session is None:
        session = Session()

    for zone in ENTSOE_ZONES:
        try:
            zone_capacity = fetch_production_capacity(zone, target_datetime, session)
            capacity_dict[zone] = zone_capacity
        except Exception:
            logger.warning(
                f"Failed to update capacity for {zone} on {target_datetime.date()}"
            )
            continue
    return capacity_dict


if __name__ == "__main__":
    fetch_production_capacity("FR", datetime(2023, 1, 1), Session())
