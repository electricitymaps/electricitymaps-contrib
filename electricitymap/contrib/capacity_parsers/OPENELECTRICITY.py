from datetime import datetime
from logging import getLogger
from typing import Any

import pandas as pd
from requests import Response, Session

from electricitymap.contrib.config import ZoneKey
from parsers.lib.utils import get_token
from parsers.OPENNEM import ZONE_KEY_TO_REGION

"""Disclaimer: only works for real-time data. There is retired capacity included but we do not have the information on when the capacity was retired."""
logger = getLogger(__name__)
REGION_MAPPING = {
    ZONE_KEY_TO_REGION[key]: key for key in ZONE_KEY_TO_REGION
}  # NT only has solar capacity so it will be excluded

FUEL_MAPPING = {
    "wind": "wind",
    "solar_rooftop": "solar",
    "battery_charging": "battery storage",
    "solar_utility": "solar",
    "coal_black": "coal",
    "battery_discharging": "battery storage",
    "pumps": "hydro storage",
    "gas_steam": "gas",
    "gas_ocgt": "gas",
    "hydro": "hydro",
    "coal_brown": "coal",
    "distillate": "oil",
    "bioenergy_biogas": "biomass",
    "gas_ccgt": "gas",
    "gas_wcmg": "gas",
    "gas_recip": "gas",
    "bioenergy_biomass": "biomass",
}

CAPACITY_URL = "https://api.openelectricity.org.au/v4/facilities/"
SOURCE = "https://openelectricity.org.au/"


def get_openelectricity_capacity_data(session: Session) -> pd.DataFrame:
    token = get_token("OPENELECTRICITY_TOKEN")
    headers = {"Authorization": f"Bearer {token}"}
    r: Response = session.request("GET", CAPACITY_URL, headers=headers)
    data = r.json()
    result = []

    time_stamp = data.get("created_at")

    for entry in data.get("data", []):
        region = entry.get("network_region")

        for unit in entry.get("units", []):
            if unit.get("dispatch_type") != "GENERATOR":
                continue

            if unit.get("status_id") != "operating":
                continue

            transformed = {
                "zone_key": region,
                "value": unit.get("capacity_registered"),
                "mode": unit.get("fueltech_id"),
                "datetime": time_stamp,
            }
            result.append(transformed)

    capacity_df = pd.DataFrame(result)
    capacity_df["datetime"] = capacity_df["datetime"].apply(
        lambda x: pd.to_datetime(x).replace(
            month=1, day=1, hour=0, minute=0, second=0, microsecond=0, tzinfo=None
        )
    )

    return capacity_df


def filter_capacity_data_by_datetime(
    data: pd.DataFrame, target_datetime: datetime
) -> pd.DataFrame:
    """Filter capacity data by datetime. For a given target_datetime, the capacity should only include plants created before then."""
    df = data.copy()
    max_datetime = df["datetime"].max()
    min_datetime = df["datetime"].min()

    if target_datetime >= max_datetime:
        df = df.copy()
    elif target_datetime <= min_datetime:
        df = df.loc[
            df["datetime"] == min_datetime
        ].copy()  # we backfill the capacity data using the first data point
    else:
        df = df.loc[df["datetime"] <= target_datetime]
    return df


def fetch_production_capacity_for_all_zones(
    target_datetime: datetime, session: Session | None = None
) -> dict[str, Any]:
    session = session or Session()
    capacity_df = get_openelectricity_capacity_data(session)
    capacity_df = filter_capacity_data_by_datetime(capacity_df, target_datetime)

    capacity_df["zone_key"] = capacity_df["zone_key"].map(REGION_MAPPING)
    capacity_df["mode"] = capacity_df["mode"].map(FUEL_MAPPING)

    capacity_df = (
        capacity_df.groupby(["zone_key", "mode"])[["value"]].sum().reset_index()
    )

    capacity_df = capacity_df.dropna(subset=["value"])

    capacity = {}
    for zone in capacity_df["zone_key"].unique():
        zone_capacity_df = capacity_df.loc[capacity_df["zone_key"] == zone]
        zone_capacity = {}
        for _idx, data in zone_capacity_df.iterrows():
            zone_capacity[data["mode"]] = {
                "datetime": target_datetime.strftime("%Y-%m-%d"),
                "value": round(data["value"], 2),
                "source": SOURCE,
            }
        capacity[zone] = zone_capacity
    return capacity


def fetch_production_capacity(
    zone_key: ZoneKey, target_datetime: datetime, session: Session | None = None
) -> dict[str, Any] | None:
    session = session or Session()
    capacity = fetch_production_capacity_for_all_zones(target_datetime, session)[
        zone_key
    ]
    if capacity:
        logger.info(
            f"Updated capacity for {zone_key} in {target_datetime}: \n{capacity}"
        )
        return capacity
    else:
        logger.error(f"No capacity data for {zone_key} in {target_datetime}")


def get_solar_capacity_au_nt(target_datetime: datetime) -> float | None:
    """Get solar capacity for AU-NT."""
    session = Session()
    capacity_df = get_openelectricity_capacity_data(session)
    capacity_df = filter_capacity_data_by_datetime(capacity_df, target_datetime)

    capacity_df = capacity_df.loc[capacity_df["zone_key"] == "NT1"]
    capacity_df["zone_key"] = "AU-NT"

    capacity_df["mode"] = capacity_df["mode"].map(FUEL_MAPPING)

    capacity_df = (
        capacity_df.groupby(["zone_key", "mode"])[["value"]].sum().reset_index()
    )

    solar_capacity = capacity_df.get("value")
    if solar_capacity is not None:
        return round(solar_capacity.values[0], 0)
    else:
        logger.error(f"No capacity data for AU-NT in {target_datetime.date()}")


if __name__ == "__main__":
    session = Session()

    print(fetch_production_capacity("AU-QLD", datetime(2025, 1, 1), session))
    # print(fetch_production_capacity("AU-VIC", datetime(2015, 1, 1), Session()))
    # print(get_solar_capacity_au_nt(datetime(2021, 1, 1)))
