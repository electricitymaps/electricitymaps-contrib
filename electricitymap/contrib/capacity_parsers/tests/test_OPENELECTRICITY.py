import os
from datetime import datetime
from pathlib import Path

import pandas as pd
import pytest
from requests_mock import ANY, GET

from electricitymap.contrib.capacity_parsers.OPENELECTRICITY import (
    fetch_production_capacity,
    filter_capacity_data_by_datetime,
)
from electricitymap.contrib.config import ZoneKey

base_path_to_mock = Path(
    "electricitymap/contrib/capacity_parsers/tests/mocks/OPENELECTRICITY"
)


test_df = pd.DataFrame(
    [
        {"datetime": datetime(2021, 1, 1), "value": 1},
        {"datetime": datetime(2022, 1, 20), "value": 1},
        {"datetime": datetime(2022, 4, 12), "value": 1},
        {"datetime": datetime(2022, 10, 3), "value": 1},
        {"datetime": datetime(2023, 3, 1), "value": 1},
    ]
)


@pytest.fixture(autouse=True)
def openelectricity_token_env():
    os.environ["OPENELECTRICITY_TOKEN"] = "token"


def test_filter_capacity_data_by_datetime():
    target_datetime_1 = datetime(2022, 3, 1)
    filtered_df_1 = filter_capacity_data_by_datetime(test_df, target_datetime_1)
    target_datetime_2 = datetime(2023, 10, 1)
    filtered_df_2 = filter_capacity_data_by_datetime(test_df, target_datetime_2)
    target_datetime_3 = datetime(2020, 10, 1)
    filtered_df_3 = filter_capacity_data_by_datetime(test_df, target_datetime_3)

    assert filtered_df_1.equals(test_df.loc[test_df["datetime"] <= target_datetime_1])
    assert filtered_df_2.equals(test_df)
    assert filtered_df_3.equals(
        test_df.loc[test_df["datetime"] <= datetime(2021, 1, 1)]
    )


def test_fetch_capacities(adapter, session, snapshot):
    data = Path(base_path_to_mock, "AU-QLD_capacities.json")
    adapter.register_uri(
        GET,
        ANY,
        text=data.read_text(),
    )

    target_datetime = datetime(2025, 1, 1)
    assert snapshot == fetch_production_capacity(
        ZoneKey("AU-QLD"), target_datetime, session
    )
