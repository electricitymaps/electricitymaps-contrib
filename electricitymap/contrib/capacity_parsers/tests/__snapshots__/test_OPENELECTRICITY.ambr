# serializer version: 1
# name: test_fetch_capacities
  dict({
    'battery storage': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 1250.6,
    }),
    'biomass': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 313.28,
    }),
    'coal': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 8119.0,
    }),
    'gas': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 3201.98,
    }),
    'hydro': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 726.0,
    }),
    'hydro storage': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 480.0,
    }),
    'oil': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 453.0,
    }),
    'solar': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 4364.81,
    }),
    'wind': dict({
      'datetime': '2025-01-01',
      'source': 'https://openelectricity.org.au/',
      'value': 2648.2,
    }),
  })
# ---
